### PR TITLE
testing: execute all tests in go submodules

### PR DESCRIPTION
**Pull request by grayside**
Part of #1180 

This takes an incremental improvement approach to the following actions:

* Front-load more of the static analysis checks.
* Perform all static analysis checks before claiming a project, this should result in less project locks overall.
* Support `gofmt` checks in submodules
* Support `go test` runs in submodules
* Sprinkles some echo statements and reduces the `set -x` output a bit to improve console readability. (Not sure if I have the right mix yet)
* Fixed assignment of `STORAGE_HMAC_ACCESS_KEY_ID` and `STORAGE_HMAC_ACCESS_SECRET_KEY` which was attempting to subshell and execute a command instead of doing string concatenation.
* Added some docs to top of file and clarifying comments.

### CODE REVIEW

**Comment by grayside** [comment]
Ah, this needs a change, else the test runs against master will double vet and double fmt the root module.

**Comment by grayside** [comment]
Updated. Note that because the CI pipeline itself lacks test coverage the lights coming green on this don't mean much. Should I temporarily add some spurious changes around the repo to see how they're handled?

**Comment by grayside** [comment]
Looks like the failure is related to #1079

File: testing/kokoro/system_tests.sh [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-# Copyright 2019 Google LLC
```
Comment (tbpg) [inline testing/kokoro/system_tests.sh]: Don't update.

File: testing/kokoro/system_tests.sh [hunk 2]
```diff
@@ -14,42 +14,141 @@
...
+## Static Analysis
+# Do the easy stuff before running tests. Fail fast!
+set +x
 
 # Fail if a dependency was added without the necessary go.mod/go.sum change
 # being part of the commit.
 # Do this before reserving a project since this doens't need a project.
-for i in `find . -name go.mod`; do
-  pushd `dirname $i` > /dev/null;
+for i in $GO_CHANGED_MODULES; do
```
Comment (tbpg) [inline testing/kokoro/system_tests.sh]: Nice. Will be good to only run this in changed directories.

File: testing/kokoro/system_tests.sh [hunk 3]
```diff
@@ -14,42 +14,141 @@
...
+
+  # Generate a list of all go files not inside a go submodule.
+  # This is necessary because go vet throws an error if you run it and no files
+  # are found to process.
+  set -x
+  files=$(find $TARGET_DIRS \( -exec [ -f {}/go.mod ] \; -prune \) -o -name "*.go" -print)
+  # Display the list of files to facilitate troubleshooting.
+  # If find has anything go vet will be run.
+  # Risk: Any unexpected find output could falsely trigger go vet.
+  # echo "$files"
```
Comment (tbpg) [inline testing/kokoro/system_tests.sh]: Delete?

File: testing/kokoro/system_tests.sh [hunk 4]
```diff
@@ -97,75 +194,21 @@ if [[ $KOKORO_BUILD_ARTIFACTS_SUBDIR = *"system-tests"* ]]; then
...
+  GO_TEST_TARGET="./..."
   echo "Running all tests"
 elif [[ -z "${TARGET_DIRS// }" ]]; then
-  TARGET=""
+  GO_TEST_TARGET=""
   echo "Only running root tests"
 else
-  TARGET=$(printf "./%s/... " $TARGET_DIRS)
+  GO_TEST_TARGET=$(printf "./%s/... " $TARGET_DIRS)
   echo "Running tests in modified directories: $TARGET"
```
Comment (tbpg) [inline testing/kokoro/system_tests.sh]: Update.

File: testing/kokoro/system_tests.sh [hunk 5]
```diff
@@ -97,75 +194,21 @@ if [[ $KOKORO_BUILD_ARTIFACTS_SUBDIR = *"system-tests"* ]]; then
...
-    go vet ./...
-    popd
-  done
-fi
-
-date
-
 OUTFILE=gotest.out
-2>&1 go test -timeout $TIMEOUT -v . $TARGET | tee $OUTFILE
+2>&1 go test -timeout $TIMEOUT -v . $GO_TEST_TARGET | tee $OUTFILE
```
Comment (tbpg) [inline testing/kokoro/system_tests.sh]: Does this end up with a single output file for all of the tests in all of the modules?

Comment (grayside) [reply testing/kokoro/system_tests.sh]: This change doesn't address submodule recursion for `go test`, my plan was to do that in a follow-up but I can add that to this PR.

Comment (grayside) [reply testing/kokoro/system_tests.sh]: I have now added test support. Please look again.

File: testing/kokoro/system_tests.sh (hunk 4 above)
Comment (grayside) [reply testing/kokoro/system_tests.sh]: Done.

File: testing/kokoro/system_tests.sh (hunk 3 above)
Comment (grayside) [reply testing/kokoro/system_tests.sh]: Done

File: testing/kokoro/system_tests.sh (hunk 1 above)
Comment (grayside) [reply testing/kokoro/system_tests.sh]: Done

**Comment by grayside** [comment]
I've added `go test` to this PR. The approach is to append the output of each test run to the same file. go-junit-report processes that into something like this:

```
<?xml version="1.0" encoding="UTF-8"?>
<testsuites>
	<testsuite tests="4" failures="0" time="0.000" name="github.com/GoogleCloudPlatform/golang-samples/run/system_package">
		<properties>
			<property name="go.version" value="go1.13.4"></property>
		</properties>
		<testcase classname="system_package" name="TestDiagramService" time="0.000">
			<skipped message="testutil.go:72: GOLANG_SAMPLES_E2E_TEST not set"></skipped>
		</testcase>
		<testcase classname="system_package" name="TestDiagramHandlerErrors" time="0.000">
			<skipped message="graphviz_test.go:111: os.Stat: stat /usr/bin/dot: no such file or directory (install graphviz?)"></skipped>
		</testcase>
		<testcase classname="system_package" name="TestDiagramHandlerImage" time="0.000">
			<skipped message="graphviz_test.go:111: os.Stat: stat /usr/bin/dot: no such file or directory (install graphviz?)"></skipped>
		</testcase>
		<testcase classname="system_package" name="TestDiagramLibrary" time="0.000">
			<skipped message="graphviz_test.go:111: os.Stat: stat /usr/bin/dot: no such file or directory (install graphviz?)"></skipped>
		</testcase>
	</testsuite>
	<testsuite tests="1" failures="0" time="0.000" name="github.com/GoogleCloudPlatform/golang-samples/run/logging-manual">
		<properties>
			<property name="go.version" value="go1.13.4"></property>
		</properties>
		<testcase classname="logging-manual" name="TestIndexHandler" time="0.000"></testcase>
	</testsuite>
	<testsuite tests="1" failures="0" time="0.000" name="github.com/GoogleCloudPlatform/golang-samples/run/grpc-ping">
		<properties>
			<property name="go.version" value="go1.13.4"></property>
		</properties>
		<testcase classname="grpc-ping" name="TestGRPCPingService" time="0.000">
			<skipped message="testutil.go:72: GOLANG_SAMPLES_E2E_TEST not set"></skipped>
		</testcase>
	</testsuite>
	<testsuite tests="2" failures="0" time="0.000" name="github.com/GoogleCloudPlatform/golang-samples/run/pubsub">
		<properties>
			<property name="go.version" value="go1.13.4"></property>
		</properties>
		<testcase classname="pubsub" name="TestHelloPubSubErrors" time="0.000"></testcase>
		<testcase classname="pubsub" name="TestHelloPubSub" time="0.000"></testcase>
	</testsuite>
	<testsuite tests="3" failures="0" time="0.000" name="github.com/GoogleCloudPlatform/golang-samples/run/hello-broken">
		<properties>
			<property name="go.version" value="go1.13.4"></property>
		</properties>
		<testcase classname="hello-broken" name="TestBrokenErrors" time="0.000"></testcase>
		<testcase classname="hello-broken" name="TestBrokenHandler" time="0.000"></testcase>
		<testcase classname="hello-broken" name="TestImprovedHandler" time="0.000"></testcase>
	</testsuite>
	<testsuite tests="1" failures="1" time="0.012" name="github.com/GoogleCloudPlatform/golang-samples/run/image-processing">
		<properties>
			<property name="go.version" value="go1.13.4"></property>
		</properties>
		<testcase classname="image-processing" name="Failure" time="0.000">
			<failure message="Failed" type="">2020/01/31 21:49:00 storage.NewClient: dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.</failure>
		</testcase>
	</testsuite>
	<testsuite tests="1" failures="1" time="0.012" name="github.com/GoogleCloudPlatform/golang-samples/run/image-processing/imagemagick">
		<properties>
			<property name="go.version" value="go1.13.4"></property>
		</properties>
		<testcase classname="imagemagick" name="Failure" time="0.000">
			<failure message="Failed" type="">2020/01/31 21:49:00 storage.NewClient: dialing: google: could not find default credentials. See https://developers.google.com/accounts/docs/application-default-credentials for more information.</failure>
		</testcase>
	</testsuite>
</testsuites>
```

**Comment by grayside** [comment]
Current test failure is because I added a few tweaks from local testing, some of which violate gofmt.

**Comment by grayside** [comment]
Here are the failing tests:

1. Cloud Functions OCR

This sample was failing but has not been under test until this PR. Looking the result, the problem is the received output includes a french ligature not used in the expected value. I've corrected that and we'll see if it fixes.

```
--- FAIL: TestDetectText (2.00s)
    ocr_test.go:184: got "2020/02/04 19:20:44 Looking for text in image menu.jpg\n2020/02/04 19:20:45 Extracted text \"suttery\\nDéjeuner du 11 Avril 1944\\nPalourdes\\nHuitres\\n---\\nRadis\\nSaucisson\\nOeufs Mimosa\\nPâté\\nFilets Merlus Sauce Verdie\\nBouchées Mongla\\nFilets de Bœuf Bristol\\nGalantine\\nPoulets de Grain Truffés\\nHaricots Maître d'Hôtel\\nSalade Ceur de Laitue\\nFromage\\nGâteaux\\nGlace\\nMoka\\nDuchesse\\n---\\nBordeaux Blanc et Rouge\\nChampagne\\nCafé\\nFine\\n\" from image (331 chars).\n2020/02/04 19:20:45 Detected language \"fr\" for text \"suttery\\nDéjeuner du 11 Avril 1944\\nPalourdes\\nHuitres\\n---\\nRadis\\nSaucisson\\nOeufs Mimosa\\nPâté\\nFilets Merlus Sauce Verdie\\nBouchées Mongla\\nFilets de Bœuf Bristol\\nGalantine\\nPoulets de Grain Truffés\\nHaricots Maître d'Hôtel\\nSalade Ceur de Laitue\\nFromage\\nGâteaux\\nGlace\\nMoka\\nDuchesse\\n---\\nBordeaux Blanc et Rouge\\nChampagne\\nCafé\\nFine\\n\".\n", want "Filets de Boeuf"
FAIL
FAIL	github.com/GoogleCloudPlatform/golang-samples/functions/ocr/app	2.931s
FAIL
```

2. GCS S3

This appears to be an existing failure to be fixed by #1079 

```
--- FAIL: TestList (0.07s)
    list_gcs_buckets_test.go:45: listGCSBuckets: MalformedSecurityHeader: Your request has a malformed header.
        	status code: 400, request id: , host id:
    list_gcs_buckets_test.go:50: listGCSBuckets got
        ----

        ----
        Want to contain
        ----
        Buckets:
        ----
FAIL
FAIL	github.com/GoogleCloudPlatform/golang-samples/storage/s3_sdk	0.080s
```

**Comment by grayside** [comment]
This is now as fixed as this PR gets. Note this might break nightly builds.

**Pull request merged**